File: Example/Example ObjC/ResultTableViewController.m
//
//  ResultTableViewController.m
//  GiniVisionExampleObjC
//
//  Created by Peter Pult on 12/08/16.
//  Copyright © 2016 Gini. All rights reserved.
//

#import "ResultTableViewController.h"
#import "AppDelegate.h"
#import <Gini-Swift.h>

@interface ResultTableViewController () {
    NSArray *_sortedKeys;
}

@end

@implementation ResultTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (_result) {
        _sortedKeys = [[_result allKeys] sortedArrayUsingSelector: @selector(compare:)];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // If a valid document is set, send feedback on it.
    // This is just to show case how to give feedback using the Gini SDK for iOS.
    // In a real world application feedback should be triggered after the user has evaluated and eventually corrected the extractions.
    [self sendFeedback:self.result];
}

- (void)sendFeedback:(NSDictionary<NSString *, Extraction *> *)result {
    
    /*******************************************
     * SEND FEEDBACK WITH THE GINI SDK FOR IOS *
     *******************************************/
    
    // As an example will set the BIC value statically.
    // In a real world application the user input should be used as the new value.
    // Feedback should only be send for labels which the user has seen. Unseen labels should be filtered out.
    
    NSString *bicValue = @"BYLADEM1001";
    Extraction *bic = result[@"bic"];
    
    NSMutableDictionary *updatedResult = [result mutableCopy];
    
    // Update or add the new value.
    if (bic) {
        bic.value = bicValue;
    } else {
        bic = [[Extraction alloc] initWithBox:nil
                                   candidates:nil
                                       entity:@"bic"
                                        value:bicValue
                                         name:@"bic"];
        updatedResult[@"bic"] = bic;
    }
    // Repeat this step for all altered fields.
    self.sendFeedbackBlock(updatedResult);
}


// MARK: Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _result.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"resultCell" forIndexPath:indexPath];
    NSString *key = _sortedKeys[indexPath.row];
        
    cell.textLabel.text = _result[key].value;
    cell.detailTextLabel.text = key;
    return cell;
}

@end

File: Example/Example Swift/ResultTableViewController.swift
//
//  ResultTableViewController.swift
//  GiniVision
//
//  Created by Peter Pult on 22/08/2016.
//  Copyright © 2016 Gini. All rights reserved.
//

import UIKit
import Gini
import GiniVision

/**
 Presents a dictionary of results from the analysis process in a table view.
 Values from the dictionary will be used as the cells titles and keys as the cells subtitles.
 */
final class ResultTableViewController: UITableViewController {
    
    /**
     The result from the analysis process.
     */
    var result: AnalysisResult? {
        
        didSet {
            let specificExtractions = result?.extractions.map({ $0.value }).sorted(by: { ($0.name ?? "") < ($1.name ?? "") }) ?? []
            let lineItems = result?.lineItems ?? []
            
            model = [("Specific Extractions", specificExtractions)]
            
            var index = 1
            
            for lineItem in lineItems {
                model += [("Line Item \(index)", lineItem.sorted(by: { ($0.name ?? "") < ($1.name ?? "") }))]
                index += 1
            }
            
            tableView.reloadData()
        }
    }
    
    private var model: [(String, [Extraction])] = []
}

extension ResultTableViewController {
    override func numberOfSections(in tableView: UITableView) -> Int {
        return model.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return model[section].1.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let extraction = model[indexPath.section].1[indexPath.item]
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "resultCell", for: indexPath)
        cell.textLabel?.text = extraction.value
        cell.detailTextLabel?.text = extraction.name
        return cell
        
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        return model[section].0
    }
}
